### PR TITLE
Fix #3796, #3786: Implement UTF-8 support in java.util.zip classes

### LINKED ISSUE #937: UTF-8 support in ZipEntry is incomplete
This is a limitation in Harmony's code, from which we ported the implementation of `ZipEntry`.

See:
- https://github.com/apache/harmony/blob/java6/classlib/modules/archive/src/main/java/java/util/zip/ZipEntry.java#L404-L411
- https://github.com/scala-native/scala-native/pull/935/files/d005de54824065ddc45490ef00730ada498cb0f0#diff-f84a19b2568acdb2532fac72a032f3c6R245

### LINKED ISSUE #3786: javalib ZipOutputStream throws exception when writing zero entries
javalib `ZipOutputStream` throws `java.util.zip.ZipException: No entries` when finishing/closing a stream
which contains zero entries.  Normally this would be a reasonable thing to do, perhaps expected. 
In this case the benchmark JVM is silent, similar to how it is silent in various finish/close cases where
one would normally expect an error.

Later:  This is pretty low in priority but I do not want to forget it.

### LINKED ISSUE #3796: javalib Zip methods need to be able to read & write UTF-8 names and comments.
This is a duplicate of PR #937 opened August 15, 2017.

The code is clearly marked as `TODO: add correct UTF-8 support.`  

From tracing so far:

* it looks like one must be careful to check .zip both a zip file written on Window and one written on
   unix-like.  The name & comment encodings may differ. 

* it looks like the SN `java.util.zip` code makes some attempt to handle two and three byte UTF-8
   but not four byte.  The code was probably written well before Java handled 4 byte UTF-8.

   I suggest replacing that code with the current practice of using a `StandardCharsets.UTF_8` when
   the zip general purpose flag indicates UTF-8 output for names & comments.

* Current practice in the SN code seems to be to always write UTF-8, even if the general purpose flag
   is clear, indicating MS-DOS CP 437 should be used. Since CP 437 is also single byte, the practice, in some places,
   in the code of using Latin-1 `iso-8859-1` should give byte compatible output, without the possibility of
   tripping over Unicode Surrogates, pairs of CP 437 can even express such.   

    The use of Latin-1 is a bit of slight-of-hand which should be commented in the file.  Exceedingly hard
    to trace and convince oneself that it is correct.  A good exam or job-interview question for those
    who thrive on `CharSets`.  CP 437 is not one of the standard Java CharSets. The slight-of-hand 
    binary passthrough   avoids some poor, benighted SN devo having to implement a custom 
   CharSet provider for it. 

<hr \hr>
Sometime in 2006 or so, the info-zip specification was updated to allow UTF-8 file names and comments.

Currently `java.util.zip` are, unconditionally, reading & writing something akin to Latin-1.  Java uses UFT-16
for characters, which is a superset of Latin-1.  That means that there are names and comments which
can be expressed in Java but not in `java.util.zip`.  

Where this may beecome  more than a theoretical gee-whiz concern is with efforts to implement  the `jdk.zipfs`
file system.

### LINKED ISSUE #3798: Warning regarding the use of labs instead of llabs
Compiling an application with Commix on Windows gives me the following warning on 0.5.0-RC1:

```
warning: absolute value function 'labs' given an argument of type 'long long' but has parameter of type 'long' which may cause truncation of value [-Wabsolute-value]
     size_t rangeSize = labs(stackBottom - stackTop);
```

(same for `size_t size = labs(it->range.address_high - it->range.address_low);`

Maybe one should use `llabs` on 64-bit systems here just in case?

### PR DESCRIPTION
**Pull request by LeeTibbert**
Fix #3796 Fix #3786 Fix #937

* javalib `java.util.zip` classes now support writing and reading UTF-8 ("Unicode Transformation Format – 8-bit")
   entry names and archive and entry comments.

* `java.util.zip.ZipOutputStream` now follows the JVM practice of not throwing an Exception is zero entries
    are written. Former behavior was sensible, but not the JVM way.
 
* both now use standard `java.lang.String` methods to do Charset conversions.  In particular, this
   should now handle 4-byte UTF-8 codepoints. 

* Scala Native `java.util.zip` methods are still limited to original zip format. No zip64.
   Java 8 and above support zip64, so there is room for improvement here.

It should be noted that .zip files have a lot of time honored complexity, both at the zip level and, especially,
with .zip files written on one operating system being readable on another.  The support implemented
in this is designed to match the JVM behavior.  A file written by JVM ought to be readable by this code
and so on for the various 2x2 combinations.

Differing operating systems may or may not be able to display the UTF-8 file and comments of this
PR. The intention is that extracting files from  archives created by the code of this PR should succeed
and have the expected UTF-8 names, to the greatest extent feasible.

TL; DR - When using UTF-8 names, if it works for you, great! If not, sorry. It may be a Scala Native bug
               or it may be the joys of zip and interoperability. The goal is to reduce the former without
               diminishing the latter.

Scala Native `java.lang` currently supports Unicode 13.0.  The September 2023 Unicode version is 15.1.
The very latest emojis, etc may not be available.  That is an open question.

### CODE REVIEW

**Comment by LeeTibbert** [comment]
Ready for review, when its turn comes around.  Thank you.

The two failures are in macOS JVM compliance.  One is a "signal 4". The other 
is an error in "pipedOutput".   Gratifying but strange that all the straight
macOS tests pass just fine.  Go figure.  Is there something different
in the environment of those two sets of tests?
